pkg/validate/serialize.go

```diff
@@ -31,14 +51,94 @@ func Verify(yamlFileName string) error {
 // (1) the yaml file can not be converted to json.
 // (2) there is a problem while unmarshalling in go type.
 // Returns an object of type olm.ClusterServiceVersion.
-func unmarshal(rawYAML []byte) (*olm.ClusterServiceVersion, error) {
+func unmarshal(rawYAML []byte) (olm.ClusterServiceVersion, error) {
 
 	var csv olm.ClusterServiceVersion
 
-	if err := yaml.Unmarshal(rawYAML, &csv); err != nil {
-		return nil, fmt.Errorf("error parsing CSV list (JSON) : %s", err)
+	rawJson, err := yaml.YAMLToJSON(rawYAML)
+	if err != nil {
+		fmt.Printf("error parsing raw YAML to Json: %s", err)
+		return csv, err
 	}
+	if err := json.Unmarshal(rawJson, &csv); err != nil {
+		return csv, fmt.Errorf("error parsing CSV list (JSON) : %s", err)
+	}
+
+	return csv, nil
+}
+
+func csvInspect(val interface{}) []string {
 
-	return &csv, nil
+	fieldValue := reflect.ValueOf(val)
+
+	switch fieldValue.Kind() {
+	case reflect.Struct:
+		return checkMissingFields(fieldValue, []string{})
+	default:
+		return []string{"Error: input file is not a valid CSV."}
+	}
+}
 
+// Takes in a string slice and checks if a string (x) is present in the slice.
+// Return true if the string is present in the slicec.
+func containsStrict(a []string, x string) bool {
+	for _, n := range a {
+		if x == n {
+			return true
+		}
+	}
+	return false
+}
+
+// Recursive function that traverses a nested struct passed in as reflect value, and reports for errors/warnings
+// in case of null struct field values.
+// Returns a log of errors as slice of strings.
+func checkMissingFields(v reflect.Value, log []string) []string {
+
+	for i := 0; i < v.NumField(); i++ {
+
+		fieldValue := v.Field(i)
+
+		tag := v.Type().Field(i).Tag.Get("json")
+		// Ignore fields that are subsets of a primitive field.
+		if tag == "" {
+			continue
+		}
+
+		fields := strings.Split(tag, ",")
+		omitEmptyTagPresent := containsStrict(fields, "omitempty")
+
+		switch fieldValue.Kind() {
+		case reflect.Struct:
+			log = checkMissingFields(fieldValue, log)
+		default:
+			emptyVal := isEmptyValue(fieldValue)
+			if emptyVal && omitEmptyTagPresent {
+				log = append(log, fmt.Sprintf("WARNING: Optional field missing (%s).", v.Type().Field(i).Name))
+			} else if emptyVal && !omitEmptyTagPresent {
+				log = append(log, fmt.Sprintf("ERROR: Mandatory field missing (%s).", v.Type().Field(i).Name))
+			}
+		}
+	}
+
+	return log
+}
+
+// Uses reflect package to check if the value of the object passed is null, returns a boolean accordingly.
+func isEmptyValue(v reflect.Value) bool {
+
+	switch v.Kind() {
+	case reflect.Array, reflect.Map, reflect.Slice, reflect.String:
+		return v.Len() == 0
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
```

Comment:
This seems wrong. What if the value is actually 0?

Reply:
Addressed this in a comment.